flash_examples/flash_components/custom_loading.py

```diff
@@ -0,0 +1,174 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import os
+from contextlib import suppress
+from typing import List, Optional
+
+from pytorch_lightning.utilities.enums import LightningEnum
+
+from flash.core.data.auto_dataset import AutoDataset
+from flash.core.data.auto_dataset_container import FlashDatasetsContainer
+from flash.core.data.data_source import DataSource, DefaultDataKeys
+from flash.core.data.utils import download_data
+from flash.core.registry import FlashRegistry
+
+download_data("https://pl-flash-data.s3.amazonaws.com/hymenoptera_data.zip", "./data")
+
+#############################################################################################
+# Imagine you have some images and they are already sorted by classes in their own folder.  #
+# You would like to create your own loading mechanism that people can re-use                #
+# Your loader would take a list of individual class folder and load the images from them    #
+# The folder paths are independent and when loading the order of folder.                    #
+# would determine the classification label.                                                 #
+# Note: This is simple enough to show you the flexibility of the Flash API.                 #
+#############################################################################################
+
+
+#############################################################################################
+#            Step 1 / 5: Create an enum to describe your new loading mechanism              #
+#############################################################################################
+
+
+class CustomDataFormat(LightningEnum):
+
+    MULTIPLE_FOLDERS = "multiple_folders"
+
+
+#############################################################################################
+#                          Step 2 / 5: Implement a DataSource.                              #
+#                                                                                           #
+# A DataSource is a state-aware (c.f training, validating, testing and predicting) dataset  #
+# and with specialized hooks (c.f load_data, load_sample) for each of those stages.         #
+# The hook resolution for the function is done in the following way.                        #
+# If {state}_load_data is implemented then it would be used exclusively for that stage.     #
+# Otherwise, it would use the load_data function.                                           #
+# `DataSource` can transform themselves into their PyTorch IterableDataset / Dataset        #
+# counterpart which is an AutoDataset/AutoIterableDataset in Flash.                         #
```

Comment:
I think I'm a little confused on the difference between a DataSource and a Dataset (they effectively have the same description)

Reply:
Yes, DataSource / AutoDataset got merged into a FlashDataset, so it is more clear.